airbyte-integrations/connectors/source-snowflake/src/main/java/io.airbyte.integrations.source.snowflake/SnowflakeDataSourceUtils.java

```diff
@@ -0,0 +1,168 @@
+    final StringBuilder jdbcUrl = new StringBuilder(
+        String.format("jdbc:snowflake://%s/?", config.get("host").asText()));
+    jdbcUrl.append(String.format(
+        "role=%s&warehouse=%s&database=%s&schema=%s&JDBC_QUERY_RESULT_FORMAT=%s&CLIENT_SESSION_KEEP_ALIVE=%s",
+        config.get("role").asText(),
+        config.get("warehouse").asText(),
+        config.get("database").asText(),
+        config.get("schema").asText(),
+        // Needed for JDK17 - see
+        // https://stackoverflow.com/questions/67409650/snowflake-jdbc-driver-internal-error-fail-to-retrieve-row-count-for-first-arrow
+        "JSON",
+        true));
+    if (config.has("jdbc_url_params")) {
+      jdbcUrl.append(config.get("jdbc_url_params").asText());
+    }
+    hikariConfig.setJdbcUrl(jdbcUrl.toString());
+
+    if (config.has("credentials") && config.get("credentials").has("auth_type")
+        && "OAuth".equals(config.get("credentials").get("auth_type").asText())) {
+      LOGGER.info("Authorization mode is OAuth");
+      try {
+        var credentials = config.get("credentials");
+        Properties properties = new Properties();
+        properties.setProperty("client_id", credentials.get("client_id").asText());
+        properties.setProperty("client_secret", credentials.get("client_secret").asText());
+        properties.setProperty("refresh_token", credentials.get("refresh_token").asText());
+        properties.setProperty("host", config.get("host").asText());
+        var accessToken = getAccessTokenUsingRefreshToken(
+            config.get("host").asText(), credentials.get("client_id").asText(),
+            credentials.get("client_secret").asText(), credentials.get("refresh_token").asText());
+        properties.put("authenticator", "oauth");
+        properties.put("token", accessToken);
+        properties.put("account", config.get("host").asText());
+        hikariConfig.setDataSourceProperties(properties);
+        // thread to keep the refresh token up to date
+        SnowflakeSource.SCHEDULED_EXECUTOR_SERVICE.scheduleAtFixedRate(getRefreshTokenTask(hikariConfig),
+            PAUSE_BETWEEN_TOKEN_REFRESH_MIN, PAUSE_BETWEEN_TOKEN_REFRESH_MIN, TimeUnit.MINUTES);
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+    } else if (config.has("credentials") && config.get("credentials").has("password")) {
+      LOGGER.info("Authorization mode is 'Username and password'");
+      var credentials = config.get("credentials");
+      hikariConfig.setUsername(credentials.get("username").asText());
+      hikariConfig.setPassword(credentials.get("password").asText());
+    } else if (config.has("password") && config.has("username")) {
+      LOGGER.info("Authorization mode is deprecated 'Username and password'. Please update your source configuration");
+      hikariConfig.setUsername(config.get("username").asText());
+      hikariConfig.setPassword(config.get("password").asText());
```

Comment:
it looks similar to the code above, can you simplify it?

Reply:
Moved to separate method